lib/cn/cn_tree_cursor.c

```diff
@@ -0,0 +1,774 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ *
+ * The cn cursor provides all keys that match the cursor's filters (prefix, view seqno etc.) in
+ * lexicographical order - forward or reverse.
+ *
+ * At any given time, a cursor only needs to merge keys across the root node and one leaf node.
+ * When enough keys are read and all keys from the leaf node are exhausted, the cursor needs to
+ * swap out the leaf node for the next leaf node in the iteration order. This is achieved using
+ * two layers of bin heaps.
+ *
+ * Each node (root node and the one leaf node) uses a binheap to merge keys across all its kvsets.
+ * These two binheaps feed into a top level binheap that forms the cn cursor.
+ *
+ * Each node uses a struct cn_level_cursor to iterate over its kvsets.
+ */
+
+#include <hse_util/event_counter.h>
+#include <hse_util/page.h>
+#include <hse_util/logging.h>
+#include <hse_util/assert.h>
+#include <hse_util/table.h>
+#include <hse_util/key_util.h>
+#include <hse_util/keycmp.h>
+#include <hse_util/bin_heap.h>
+
+#include <hse_ikvdb/cn.h>
+
+#include <cn/cn_cursor.h>
+
+#include "route.h"
+
+#define MTF_MOCK_IMPL_cn_tree_cursor
+
+#include "cn_tree_internal.h"
+#include "cn_tree_cursor.h"
+#include "kvset_internal.h"
+#include "kvset.h"
+#include "kv_iterator.h"
+
+/*
+ * Min heap comparator for forward iteration.
+ *
+ * Returns:
+ *   < 0 : a_blob < b_blob
+ *   > 0 : a_blob > b_blob
+ *  == 0 : a_blob == b_blob
+ */
+static int
+cn_kv_cmp(const void *a_blob, const void *b_blob)
+{
+    const struct cn_kv_item *a = a_blob;
+    const struct cn_kv_item *b = b_blob;
+
+    return key_obj_cmp(&a->kobj, &b->kobj);
+}
+
+/*
+ * Max heap comparator for reverse iteration.
+ * with a caveat: A ptomb sorts before all keys w/ matching
+ * prefix.
+ *
+ * Returns:
+ *   < 0 : a_blob > b_blob
+ *   > 0 : a_blob < b_blob
+ *  == 0 : a_blob == b_blob
+ */
+static int
+cn_kv_cmp_rev(const void *a_blob, const void *b_blob)
+{
+    const struct cn_kv_item *a = a_blob;
+    const struct cn_kv_item *b = b_blob;
+    size_t                   a_klen = a->kobj.ko_pfx_len + a->kobj.ko_sfx_len;
+    size_t                   b_klen = b->kobj.ko_pfx_len + b->kobj.ko_sfx_len;
+
+    int rc;
+
+    if (!(a->vctx.is_ptomb ^ b->vctx.is_ptomb))
+        return key_obj_cmp(&b->kobj, &a->kobj);
+
+    /* Exactly one of a and b is a ptomb. */
+    if (a->vctx.is_ptomb && a_klen <= b_klen) {
+        rc = key_obj_ncmp(&b->kobj, &a->kobj, a_klen);
+        if (rc == 0)
+            return -1; /* a wins */
+    } else if (b->vctx.is_ptomb && b_klen <= a_klen) {
+        rc = key_obj_ncmp(&b->kobj, &a->kobj, b_klen);
+        if (rc == 0)
+            return 1; /* b wins */
+    }
+
+    /* Non-ptomb key is shorter than ptomb. Full key compare. */
+    return key_obj_cmp(&b->kobj, &a->kobj);
+}
+
+MTF_STATIC merr_t
+cn_tree_kvset_refs(struct cn_tree_node *node, struct cn_level_cursor *lcur)
```

Comment:
Thoughts for a follow-on PR:
- This function should have "get" in its name.
- Make it more generic and more testable by passing table, dgenhi, dgenlo instead of ptr to cursor.  Go one step further and pass a ptr to the kvset list.
- Do we know the kvset list size up front?  If so, maybe allocate enough space in the table up front.
- Function comment that caller needs to have node/tree lock.